packages/flutter/lib/src/semantics/semantics.dart

```diff
@@ -1431,18 +1431,74 @@ class SemanticsNode extends AbstractNode with DiagnosticableTreeMixin {
   }
 }
 
-/// A helper class to contain a semantics node and the effective
-/// sort order of that node during the traversal of the tree. This is
-/// what is actually sorted when semantics nodes are sorted.
+/// This class defines the comparison that is used to sort [SemanticsNode]s
+/// before sending them to the platform side.
+///
+/// This is a helper class used to contain a [node], the effective
+/// [order], the globally transformed starting corner [globalStartCorner],
+/// and the containing node's [containerTextDirection] during the traversal of
+/// the semantics node tree. A null value is allowed for [containerTextDirection],
+/// because in that case we want to fall back to ordering by child insertion
+/// order for nodes that are equal after sorting from top to bottom.
 class _TraversalSortNode implements Comparable<_TraversalSortNode> {
-  _TraversalSortNode(this.node, this.order);
+  _TraversalSortNode(this.node, this.order, this.containerTextDirection, Matrix4 transform)
+    : assert(node != null) {
+    // When containerTextDirection is null, this is set to topLeft, but the x
+    // coordinate is also ignored when doing the comparison in that case, so
+    // this isn't actually expressing a directionality opinion.
+    globalStartCorner = _transformPoint(
+      containerTextDirection == TextDirection.rtl ? node.rect.topRight : node.rect.topLeft,
```

Comment:
pass in topRight when direction is null, so we're more likely to catch regressions where someone starts depending on it